RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronSCProducer.cc

```diff
@@ -0,0 +1,162 @@
+      const reco::PFRecTrackRef& kfpf = gsfpf->kfPFRecTrackRef();
+      point3 = kfpf->extrapolatedPoint(LayerType::ECALShowerMax);
+      reco::PFClusterRef best_kf = closest_cluster( point3, ecalClusters, ecal_matched );
+      if( !best_kf.isNull() ) { 
+	best_seed = best_kf; 
+	clusteredRefs.push_back(best_kf); 
+      }
+    }
+    
+    //    // get closest HCAL PF cluster
+    //    auto hcal_position = pftrk->extrapolatedPoint(reco::PFTrajectoryPoint::HCALEntrance);
+    //    PFClusterRef hcal_ref = closest_cluster(hcal_position, hcal_clusters);
+    //    if(!hcal_ref.isNull()) {
+    //      hcal_ktf_clusters_map.insert(std::pair<reco::TrackRef, PFClusterRef>(trk, hcal_ref));
+    //    }
+
+    //now we need to make the supercluster
+    if(!best_seed.isNull()) {
+
+      float posX=0.,posY=0.,posZ=0.;
+      float scEnergy=0.;
+      for(const auto clus : clusteredRefs){
+	scEnergy+=clus->correctedEnergy();
+	posX+=clus->correctedEnergy()*clus->position().X();
+	posY+=clus->correctedEnergy()*clus->position().Y();
+	posZ+=clus->correctedEnergy()*clus->position().Z();
+      }
+      posX/=scEnergy;
+      posY/=scEnergy;
+      posZ/=scEnergy;
+      reco::SuperCluster new_sc(scEnergy,math::XYZPoint(posX,posY,posZ));   
+      new_sc.setCorrectedEnergy(scEnergy);
+      new_sc.setSeed(edm::refToPtr(best_seed));
+      std::vector<const reco::PFCluster*> barePtrs;
+      for(const auto clus : clusteredRefs){
+	new_sc.addCluster(edm::refToPtr(clus));
+	barePtrs.push_back(&*clus);
+      }
+      PFClusterWidthAlgo pfwidth(barePtrs);
+      new_sc.setEtaWidth(pfwidth.pflowEtaWidth());
+      new_sc.setPhiWidth(pfwidth.pflowPhiWidth());
+      new_sc.rawEnergy();//cache the value of raw energy
+      
+      // Store new SuperCluster 
+      clusters->push_back( new_sc );
+      
+      // Populate ValueMap container
+      clusters_valuemap.push_back( reco::SuperClusterRef(clusters_refprod,igsfpf) );
+    }else{
+      clusters_valuemap.push_back( reco::SuperClusterRef(clusters_refprod.id()) );
+    }
+  }
+
+  // Put SuperClusters in event
+  event.put(std::move(clusters));
+
+  // Put ValueMap<SuperClusterRef> in event
+  std::unique_ptr< edm::ValueMap<reco::SuperClusterRef> > ptr( new edm::ValueMap<reco::SuperClusterRef>() );
+  edm::ValueMap<reco::SuperClusterRef>::Filler filler(*ptr);
+  filler.insert(gsfPfRecTracks, clusters_valuemap.begin(), clusters_valuemap.end());
+  filler.fill();
+  event.put(std::move(ptr));
+
+}
+
+reco::PFClusterRef LowPtGsfElectronSCProducer::closest_cluster( const reco::PFTrajectoryPoint& point,
+								const edm::Handle<PFClusters>& clusters,
+								std::vector<int>& matched ) {
+  reco::PFClusterRef best_ref;
+  if ( point.isValid() ) {
+    float min_dr = 1.e6;
+    for ( size_t ii = 0; ii < clusters->size(); ++ii ) {
+      float dr = reco::deltaR( clusters->at(ii), point.positionREP() );
+      if( dr < min_dr ) {// && std::find( matched.begin(), matched.end(), ii ) == matched.end() ) {//@@
```

Comment:
@Sam-Harper I do not record which clusters have been matched or not, and hence I do not check if they are used multiple times or not ... I vaguely remember that this was due to problems, but I cannot remember what exactly. Should we revisit this?

Reply:
I think we should enforce  uniqueness as it would be surprisingly.

Reply:
Done, [here](https://github.com/CMSBParking/cmssw/pull/3/commits/31fe793604399e4ba5d5480ba12da6195f0bf330#diff-378f6e137ecbdaa7fef52f166d3628dcR157)